packages/react-client/src/DeviceManager.ts

```diff
@@ -1,563 +1,163 @@
+    requestedMedia: boolean,
+    error: DeviceError | null = null,
+  ) => {
+    this.deviceState = prepareDeviceState(stream, track, devices, error, requestedMedia);
 
-    if (result.type === "Error" && result.error?.name === "NotAllowedError") {
-      result = await handleNotAllowedError(result.constraints);
-    }
-
-    const mediaDeviceInfos: MediaDeviceInfo[] = await navigator.mediaDevices.enumerateDevices();
-
-    if (result.type === "OK") {
-      requestedDevices = result.stream;
-      // Safari changes deviceId between sessions, therefore we cannot rely on deviceId for identification purposes.
-      // We can switch a random device that comes from safari to one that has the same label as the one used in the previous session.
-      const currentDevices = getCurrentDevicesSettings(requestedDevices, mediaDeviceInfos);
-      const shouldCorrectDevices = isAnyDeviceDifferentFromLastSession(
-        previousVideoDevice,
-        previousAudioDevice,
-        currentDevices,
-      );
-      if (shouldCorrectDevices) {
-        const videoIdToStart = mediaDeviceInfos.find((info) => info.label === previousVideoDevice?.label)?.deviceId;
-        const audioIdToStart = mediaDeviceInfos.find((info) => info.label === previousAudioDevice?.label)?.deviceId;
-
-        if (videoIdToStart || audioIdToStart) {
-          stopTracks(requestedDevices);
-
-          const exactConstraints: MediaStreamConstraints = {
-            video: !!result.constraints.video && prepareConstraints(videoIdToStart, videoTrackConstraints),
-            audio: !!result.constraints.video && prepareConstraints(audioIdToStart, audioTrackConstraints),
-          };
-
-          const correctedResult = await getMedia(exactConstraints, result.previousErrors);
-
-          if (correctedResult.type === "OK") {
-            requestedDevices = correctedResult.stream;
-          } else {
-            console.error("Device Manager unexpected error");
-          }
-        }
-      }
-    }
-
-    const video: DeviceState = prepareDeviceState(
-      requestedDevices,
-      requestedDevices?.getVideoTracks()[0] || null,
-      mediaDeviceInfos.filter(isVideo),
-      getError(result, "video"),
-      shouldAskForVideo,
-    );
-
-    const audio: DeviceState = prepareDeviceState(
-      requestedDevices,
-      requestedDevices?.getAudioTracks()[0] || null,
-      mediaDeviceInfos.filter(isAudio),
-      getError(result, "audio"),
-      shouldAskForAudio,
-    );
-
-    this.video = video;
-    this.audio = audio;
-
-    if (video.media?.deviceInfo) {
-      this.saveLastVideoDevice(video.media?.deviceInfo);
-    }
-
-    if (audio.media?.deviceInfo) {
-      this.saveLastAudioDevice(audio.media?.deviceInfo);
-    }
+    const deviceInfo = this.deviceState.media?.deviceInfo;
+    if (deviceInfo) this.saveLastDevice(deviceInfo);
 
+    this.status = "initialized";
     this.setupOnEndedCallback();
+    this.emit("managerInitialized", this.deviceState);
```

Comment:
Now this event is emitted twice (once for every device manager). Maybe it should be moved to `initializeDevices`